extensions/amp-ad/0.1/amp-ad-custom.js

```diff
@@ -0,0 +1,120 @@
+/**
+ * Copyright 2016 David Karlin, Bachtrack Ltd. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {isLayoutSizeDefined} from '../../../src/layout';
+import {user} from '../../../src/log';
+import {templatesFor} from '../../../src/template';
+import {xhrFor} from '../../../src/xhr';
+
+/** @const {!string} Tag name for custom ad implementation. */
+export const TAG_AD_CUSTOM = 'amp-ad-custom';
+
+/** @var {Object} A map of promises for each value of data-url. The promise
+ *  will fetch data for the URL for the ad server, and return it as a map of
+ *  objects, keyed by slot; each object contains the variables to be
+ *   substituted into the mustache template. */
+const ampCustomadXhrPromises = {};
+
+/** @var {Object} a map of ad slot ids for each value of data-url.
+ *  Each entry in the map is an array of slot ids. */
+let ampCustomadSlots = null;
```

Comment:
I think it lost my previous comments because of the file renaming.

Could you have the map to cache the final URL directly instead of caching an array of slots?